src/Storages/MergeTree/MarkRange.cpp

```diff
@@ -3,6 +3,24 @@
 namespace DB
 {
 
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+bool MarkRange::operator==(const MarkRange & rhs) const
+{
+    return begin == rhs.begin && end == rhs.end;
+}
+
+bool MarkRange::operator<(const MarkRange & rhs) const
+{
+    /// We allow only consecutive non-intersecting ranges
+    if (rhs.begin > begin && rhs.begin < end)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Bad marks");
```

Comment:
Again, I don't understand why do we skip other intersections.
Exception message gives no introspection. Print ranges there.